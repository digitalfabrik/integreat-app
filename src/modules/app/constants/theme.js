import colors from './colors'
import dimensions from './dimensions'
import fonts from './fonts'
import helpers from './helpers'

export default {
  colors,
  dimensions,
  fonts,
  helpers
}
