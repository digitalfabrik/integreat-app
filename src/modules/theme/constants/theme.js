// @flow

import buildConfig from '../../app/constants/buildConfig'
import type { ThemeType } from '../../../../build-configs/themes/ThemeType'

export type { ThemeType }

export const brightTheme: ThemeType = buildConfig.theme

export const darkTheme: ThemeType = buildConfig.darkTheme

export default brightTheme
