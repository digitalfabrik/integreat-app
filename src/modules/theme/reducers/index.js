// @flow

import { handleAction, type ReduxReducer } from 'redux-actions'
import type { ToggleDarkModeActionType } from '../../app/StoreActionType'

const toggleDarkModeReducer: ReduxReducer<boolean, ToggleDarkModeActionType> =
  handleAction(
    'TOGGLE_DARK_MODE',
    (state: boolean) => !state,
    false
  )

export default toggleDarkModeReducer
