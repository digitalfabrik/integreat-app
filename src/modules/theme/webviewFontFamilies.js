// @flow

import type { ThemeType } from './constants/theme'

const ARABIC_LANGUAGES = ['ar', 'fa', 'ku']

const webviewFontFamilies = (theme: ThemeType, language: string) => {
  return ARABIC_LANGUAGES.includes(language) ? theme.fonts.arabicWebviewFontFamilies : theme.fonts.webviewFontFamilies
}

export default webviewFontFamilies
