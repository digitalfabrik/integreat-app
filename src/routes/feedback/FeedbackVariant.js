// @flow

import type {
  FeedbackType, FeedbackCategoryType
} from '@integreat-app/integreat-api-client/endpoints/createFeedbackEndpoint'

class FeedbackVariant {
  label: string
  language: string
  city: string
  feedbackType: FeedbackType
  feedbackCategory: FeedbackCategoryType
  pagePath: ?string
  alias: ?string

  constructor (label: string, language: string, city: string, feedbackType: FeedbackType,
    feedbackCategory: FeedbackCategoryType, pagePath?: string, alias?: string) {
    this.label = label
    this.language = language
    this.city = city
    this.feedbackType = feedbackType
    this.feedbackCategory = feedbackCategory
    this.pagePath = pagePath
    this.alias = alias
  }
}

export default FeedbackVariant
