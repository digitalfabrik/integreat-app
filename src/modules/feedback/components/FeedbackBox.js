// @flow

import * as React from 'react'

import type { TFunction } from 'react-i18next'
import { withTranslation } from 'react-i18next'
import styled from 'styled-components'
import ModalHeader from './ModalHeader'
import FeedbackComment from './FeedbackComment'
import FeedbackVariant from '../FeedbackVariant'
import TextButton from '../../common/components/TextButton'
import type { ThemeType } from '../../theme/constants/theme'
import Dropdown from '../../common/components/Dropdown'
import type { SendingStatusType } from './FeedbackModal'

export const StyledFeedbackBox = styled.div`
  display: flex;
  width: 400px;
  height: auto;
  box-sizing: border-box;
  flex-direction: column;
  justify-content: space-between;
  padding: 20px;
  border-radius: 10px;
  border-color: #585858;
  font-size: ${props => props.theme.fonts.contentFontSize};
`

export const Description = styled.div`
  padding: 10px 0 5px;
`

type PropsType = {|
  isPositiveRatingSelected: boolean,
  feedbackOptions: Array<FeedbackVariant>,
  selectedFeedbackOption: FeedbackVariant,
  comment: string,
  onCommentChanged: SyntheticInputEvent<HTMLTextAreaElement> => void,
  onFeedbackOptionChanged: FeedbackVariant => void,
  onSubmit: () => void,
  t: TFunction,
  closeFeedbackModal: () => void,
  theme: ThemeType,
  sendingStatus: SendingStatusType
|}

/**
 * Renders all necessary inputs for a Feedback and posts the data to the feedback endpoint
 */
export class FeedbackBox extends React.PureComponent<PropsType> {
  render () {
    const {
      selectedFeedbackOption,
      feedbackOptions,
      t,
      isPositiveRatingSelected,
      onFeedbackOptionChanged,
      onCommentChanged,
      onSubmit,
      comment,
      closeFeedbackModal,
      theme,
      sendingStatus
    } = this.props

    return (
      <StyledFeedbackBox>
        <ModalHeader t={t} closeFeedbackModal={closeFeedbackModal} title={t('feedback')} />
        <Description>{t('feedbackType')}</Description>
        <Dropdown items={feedbackOptions}
                  selectedItem={selectedFeedbackOption}
                  onOptionChanged={onFeedbackOptionChanged}
                  theme={theme} />
        <FeedbackComment
          comment={comment}
          commentMessage={isPositiveRatingSelected ? t('positiveComment') : t('negativeComment')}
          onCommentChanged={onCommentChanged}
          required={!isPositiveRatingSelected} />
        {sendingStatus === 'ERROR' && <Description>{t('failedSendingFeedback')}</Description>}
        <TextButton
          disabled={!isPositiveRatingSelected && !comment}
          onClick={onSubmit}
          text={t('send')} />
      </StyledFeedbackBox>
    )
  }
}

export default withTranslation('feedback')(FeedbackBox)
