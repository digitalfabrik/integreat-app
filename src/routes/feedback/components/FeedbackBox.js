// @flow

import * as React from 'react'
import 'react-dropdown/style.css'

import CityModel from '../../../modules/endpoint/models/CityModel'
import { translate } from 'react-i18next'
import styled from 'styled-components'
import FeedbackEndpoint, {
  CATEGORIES_FEEDBACK_TYPE,
  DEFAULT_FEEDBACK_LANGUAGE,
  EXTRA_FEEDBACK_TYPE,
  INTEGREAT_INSTANCE,
  PAGE_FEEDBACK_TYPE,
  SEARCH_FEEDBACK_TYPE
}
  from '../../../modules/endpoint/FeedbackEndpoint'
import type { TFunction } from 'react-i18next'
import CleanLink from '../../../modules/common/components/CleanLink'
import type { FeedbackDataType } from '../../../modules/endpoint/FeedbackEndpoint'
import { CATEGORIES_ROUTE } from '../../../modules/app/routes/categories'
import { EVENTS_ROUTE } from '../../../modules/app/routes/events'
import { EXTRAS_ROUTE } from '../../../modules/app/routes/extras'
import { SEARCH_ROUTE } from '../../../modules/app/routes/search'
import { DISCLAIMER_ROUTE } from '../../../modules/app/routes/disclaimer'
import FeedbackHeader from './FeedbackHeader'
import FeedbackComment from './FeedbackComment'
import type { LocationState } from 'redux-first-router'
import { goToFeedback } from '../../../modules/app/routes/feedback'
import FeedbackDropdownItem from '../FeedbackDropdownItem'

const StyledFeedbackBox = styled.div`
  display: flex;
  width: 400px;
  height: auto;
  box-sizing: border-box;
  flex-direction: column;
  justify-content: space-between;
  padding: 20px;
  border-radius: 10px;
  border-color: #585858;
  font-size: ${props => props.theme.fonts.contentFontSize};
`

export const Description = styled.div`
  padding: 10px 0 5px;
`

const SubmitButton = styled(CleanLink)`
  margin: 15px 0;
  padding: 5px;
  background-color: ${props => props.theme.colors.themeColor};
  color: ${props => props.theme.colors.backgroundAccentColor};
  text-align: center;
  border-radius: 0.25em;
`

type PropsType = {
  cities: Array<CityModel>,
  title?: string,
  alias?: string,
  id?: number,
  query?: string,
  isPositiveRatingSelected: boolean,
  location: LocationState,
  isOpen: boolean,
  commentMessageOverride: ?string,
  hideHeader: boolean,
  t: TFunction
}

type StateType = {
  feedbackOptions: Array<FeedbackDropdownItem>,
  selectedFeedbackOption: FeedbackDropdownItem,
  comment: string
}

export class FeedbackBox extends React.Component<PropsType, StateType> {
  static defaultProps = {
    hideHeader: false
  }
  constructor (props: PropsType) {
    super(props)
    const feedbackOptions = this.getFeedbackOptions()
    this.state = {feedbackOptions: feedbackOptions, selectedFeedbackOption: feedbackOptions[0], comment: ''}
  }

  onCommentChanged = (event: SyntheticInputEvent<HTMLTextAreaElement>) => this.setState({comment: event.target.value})

  onFeedbackOptionChanged = (selectedDropdown: FeedbackDropdownItem) => {
    this.setState({selectedFeedbackOption: selectedDropdown})
  }

  componentDidUpdate (prevProps: PropsType) {
    const {isOpen} = this.props
    const prevIsOpen = prevProps.isOpen

    // If the FeedbackBox is opened, we have to reset and initialize the state
    if (prevIsOpen !== isOpen && isOpen) {
      /* eslint-disable react/no-did-update-set-state */
      const feedbackOptions = this.getFeedbackOptions()
      const selectedFeedbackOption = feedbackOptions[0]
      this.setState({feedbackOptions: feedbackOptions, selectedFeedbackOption: feedbackOptions[0], comment: ''})
      FeedbackEndpoint.postData(this.getFeedbackData(selectedFeedbackOption, ''))
    }
  }

  getFeedbackData = (selectedFeedbackOption: FeedbackDropdownItem, comment: string): FeedbackDataType => {
    const {location, query, isPositiveRatingSelected, id, alias} = this.props
    const {city, language} = location.payload
    return {
      feedbackType: selectedFeedbackOption.feedbackType,
      isPositiveRating: isPositiveRatingSelected,
      comment,
      id,
      city: city || INTEGREAT_INSTANCE,
      language: language || DEFAULT_FEEDBACK_LANGUAGE,
      alias,
      query
    }
  }

  onSubmit = () => {
    const {selectedFeedbackOption, comment} = this.state
    FeedbackEndpoint.postData(this.getFeedbackData(selectedFeedbackOption, comment))
  }

  getFeedbackOptions = (): Array<FeedbackDropdownItem> => {
    const {cities, location, t} = this.props
    const {city} = location.payload

    const options = []
    const currentPageFeedbackLabel = this.getCurrentPageFeedbackLabel()
    if (currentPageFeedbackLabel) {
      options.push(this.getFeedbackOption(currentPageFeedbackLabel, this.getCurrentPageFeedbackType()))
    }
    if (city) {
      const cityTitle = CityModel.findCityName(cities, city)
      options.push(this.getFeedbackOption(`${t('contentOfCity')} ${cityTitle}`, CATEGORIES_FEEDBACK_TYPE))
    }

    options.push(this.getFeedbackOption(t('technicalTopic'), CATEGORIES_FEEDBACK_TYPE))

    return options
  }

  getFeedbackOption = (label: string, feedbackType: ?string): FeedbackDropdownItem =>
    new FeedbackDropdownItem(label, feedbackType)

  getCurrentPageFeedbackLabel = (): ?string => {
    const {location, id, alias, title, query, t} = this.props
    const type = location.type

    if (type === CATEGORIES_ROUTE && id && title) {
      return `${t('contentOfPage')} '${title}'`
    } else if (type === EVENTS_ROUTE && id && title) {
      return `${t('news')} '${title}'`
    } else if (type === EXTRAS_ROUTE && alias && title) {
      return `${t('extra')} '${title}'`
    } else if (type === SEARCH_ROUTE && query) {
      return `${t('searchFor')} '${query}'`
    } else if (type === DISCLAIMER_ROUTE) {
      return `${t('disclaimer')}`
    } else {
      return null
    }
  }

  getCurrentPageFeedbackType = (): string | null => {
    const {type} = this.props.location
    if (type === SEARCH_ROUTE) {
      return SEARCH_FEEDBACK_TYPE
    } else if (type === EXTRAS_ROUTE) {
      return EXTRA_FEEDBACK_TYPE
    } else {
      return PAGE_FEEDBACK_TYPE
    }
  }

  render () {
    const {selectedFeedbackOption, feedbackOptions, comment} = this.state
    const {t, isPositiveRatingSelected, location, commentMessageOverride, hideHeader} = this.props
    return (
      <StyledFeedbackBox>
        {!hideHeader && (
          <FeedbackHeader
            location={location}
            selectedFeedbackOption={selectedFeedbackOption}
            feedbackOptions={feedbackOptions}
            onFeedbackOptionChanged={this.onFeedbackOptionChanged} />
        )}
        <FeedbackComment
          comment={comment}
          commentMessageOverride={commentMessageOverride}
          isPositiveRatingSelected={isPositiveRatingSelected}
          onCommentChanged={this.onCommentChanged} />
        <SubmitButton to={goToFeedback(location)} onClick={this.onSubmit}>
          {t('send')}
        </SubmitButton>
      </StyledFeedbackBox>
    )
  }
}

export default translate('feedback')(FeedbackBox)
