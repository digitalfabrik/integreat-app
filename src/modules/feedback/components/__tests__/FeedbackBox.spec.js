// @flow

import React from 'react'
import { shallow } from 'enzyme'
import { FeedbackBox } from '../FeedbackBox'
import FeedbackVariant from '../../FeedbackVariant'
import { CATEGORIES_FEEDBACK_TYPE, CONTENT_FEEDBACK_CATEGORY } from '@integreat-app/integreat-api-client'
import theme from '../../../theme/constants/theme'

describe('FeedbackBox', () => {
  const t = (key: ?string): string => key || ''
  const feedbackOptions = [
    new FeedbackVariant({
      label: 'label',
      feedbackType: CATEGORIES_FEEDBACK_TYPE,
      feedbackCategory: CONTENT_FEEDBACK_CATEGORY
    })
  ]
  const onCommentChanged = (event: SyntheticInputEvent<HTMLTextAreaElement>) => {}
  const onFeedbackOptionChanged = (option: FeedbackVariant) => {}
  const onSubmit = () => {}

  it('should match snapshot', () => {
    const component = shallow(
      <FeedbackBox
        isPositiveRatingSelected={false}
        comment=''
        feedbackOptions={feedbackOptions}
        selectedFeedbackOption={feedbackOptions[0]}
        onCommentChanged={onCommentChanged}
        onFeedbackOptionChanged={onFeedbackOptionChanged}
        onSubmit={onSubmit}
        sendingStatus='SUCCESS'
        t={t}
        closeFeedbackModal={() => {}}
        theme={theme} />
    )
    expect(component).toMatchSnapshot()
  })
})
