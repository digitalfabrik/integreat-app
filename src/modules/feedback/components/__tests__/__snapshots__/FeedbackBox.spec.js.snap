// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackBox should match snapshot 1`] = `
<FeedbackBox__StyledFeedbackBox>
  <ModalHeader
    closeFeedbackModal={[Function]}
    t={[Function]}
    title="feedback"
  />
  <FeedbackBox__Description>
    feedbackType
  </FeedbackBox__Description>
  <Dropdown
    items={
      Array [
        FeedbackVariant {
          "alias": undefined,
          "feedbackCategory": "Inhalte",
          "feedbackType": "categories",
          "label": "label",
          "value": "label",
        },
      ]
    }
    onOptionChanged={[Function]}
    selectedItem={
      FeedbackVariant {
        "alias": undefined,
        "feedbackCategory": "Inhalte",
        "feedbackType": "categories",
        "label": "label",
        "value": "label",
      }
    }
    theme={
      Object {
        "colors": Object {
          "backgroundAccentColor": "#fafafa",
          "backgroundColor": "#ffffff",
          "textColor": "#000000",
          "textDecorationColor": "#c7c7c7",
          "textDisabledColor": "#d0d0d0",
          "textSecondaryColor": "#585858",
          "themeColor": "#fbda16",
          "tunewsThemeColor": "#007aa8",
          "tunewsThemeColorLight": "rgba(0, 122, 168, 0.4)",
        },
        "dimensions": Object {
          "headerHeightLarge": 85,
          "headerHeightSmall": 55,
          "maxWidth": 1100,
          "minMaxWidth": "screen and (min-width: 1100px)",
          "smallViewport": "(max-width: 750px)",
          "toolbarWidth": 125,
        },
        "fonts": Object {
          "contentFontFamily": "'Open Sans', 'Lateef', sans-serif",
          "contentFontSize": "0.95rem",
          "contentLineHeight": 1.4,
          "decorativeFontFamily": "'Raleway', 'Lateef', sans-serif",
          "decorativeFontSize": "0.95rem",
          "decorativeLineHeight": 1.3,
          "fontSizeAdjust": 0.55,
          "standardParagraphMargin": "0.75rem",
          "subTitleFontSize": "1.8rem",
        },
        "helpers": Object {
          "removeLinkHighlighting": "
    color: inherit;
    text-decoration: none;
  ",
        },
      }
    }
  />
  <FeedbackComment
    comment=""
    commentMessage="negativeComment"
    onCommentChanged={[Function]}
    required={true}
  />
  <TextButton
    disabled={true}
    onClick={[Function]}
    text="send"
  />
</FeedbackBox__StyledFeedbackBox>
`;
